streams/src/test/java/org/apache/kafka/streams/kstream/internals/CogroupedKStreamImplTest.java

```diff
@@ -0,0 +1,483 @@
+import org.apache.kafka.test.MockProcessorSupplier;
+import org.apache.kafka.test.StreamsTestUtils;
+import org.junit.Before;
+import org.junit.Test;
+
+public class CogroupedKStreamImplTest {
+
+    private final Consumed<String, String> stringConsumed = Consumed
+            .with(Serdes.String(), Serdes.String());
+    private final MockProcessorSupplier<String, String> processorSupplier = new MockProcessorSupplier<>();
+    private static final String TOPIC = "topic";
+    private static final String OUTPUT = "output";
+    private final StreamsBuilder builder = new StreamsBuilder();
+    private KGroupedStream<String, String> groupedStream;
+    private CogroupedKStream<String, String> cogroupedStream;
+
+    private final Properties props = StreamsTestUtils
+        .getStreamsConfig(Serdes.String(), Serdes.String());
+
+    private static final Aggregator<String, String, String> STRING_AGGREGATOR = (key, value, aggregate) ->
+        aggregate + value;
+
+    private static final Initializer<String> STRING_INITIALIZER = () -> "";
+
+    private static final Aggregator<String, String, Integer> STRSUM_AGGREGATOR = (key, value, aggregate) ->
+        aggregate + Integer.parseInt(value);
+
+    private static final Aggregator<? super String, ? super Integer, Integer> SUM_AGGREGATOR = (key, value, aggregate) ->
+        aggregate + value;
+
+    private static final Initializer<Integer> SUM_INITIALIZER = () -> 0;
+
+
+    @Before
+    public void setup() {
+        final KStream<String, String> stream = builder.stream(TOPIC, Consumed
+            .with(Serdes.String(), Serdes.String()));
+
+        groupedStream = stream.groupByKey(Grouped.with(Serdes.String(), Serdes.String()));
+        cogroupedStream = groupedStream.cogroup(MockAggregator.TOSTRING_ADDER);
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void shouldThrowNPEInCogroupIfKGroupedStreamIsNull() {
+        cogroupedStream.cogroup(null, MockAggregator.TOSTRING_ADDER);
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void shouldNotHaveNullAggregatorOnCogroup() {
+        cogroupedStream.cogroup(groupedStream, null);
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void shouldNotHaveNullInitializerOnAggregate() {
+        cogroupedStream.aggregate(null, Materialized.as("store"));
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void shouldNotHaveNullMaterMaterializedOnAggregate() {
+        cogroupedStream.aggregate(STRING_INITIALIZER, (Materialized<String, String, KeyValueStore<Bytes, byte[]>>) null);
+    }
+    
+
+    @Test
+    public void useNamed() {
+
+        final KStream<String, String> test = builder.stream("one", stringConsumed);
+
+        final KGroupedStream<String, String> groupedOne = test.groupByKey();
+
+        final KTable<String, String> customers = groupedOne.cogroup(STRING_AGGREGATOR)
+                .aggregate(STRING_INITIALIZER, Named.as("test"));
```

Comment:
I was using this test to print the topology and it shows two sub topologies while it should be one (seems the reason is that you use the same `StreamsBuilder` as in `setup()` method.

Also, the naming of the operators seems to be incorrect. Also wondering if `KGroupedStream#cogroup()` needs on overload that takes a `Named` parameter? Maybe not, but the specified `Named` from `aggregate()` would need to be used for other processors, too. Atm there is this weird `COGROUPKSTREAM-AGGREGATE-KSTREAM-SOURCE-0000000001test`

```
Topologies:
   Sub-topology: 0
    Source: KSTREAM-SOURCE-0000000000 (topics: [topic])
      --> none

  Sub-topology: 1
    Source: KSTREAM-SOURCE-0000000001 (topics: [one])
      --> COGROUPKSTREAM-AGGREGATE-KSTREAM-SOURCE-0000000001test
    Processor: COGROUPKSTREAM-AGGREGATE-KSTREAM-SOURCE-0000000001test (stores: [COGROUPKSTREAM-AGGREGATE-STATE-STORE-0000000002])
      --> test
      <-- KSTREAM-SOURCE-0000000001
    Processor: test (stores: [COGROUPKSTREAM-AGGREGATE-STATE-STORE-0000000002])
      --> KTABLE-TOSTREAM-0000000005
      <-- COGROUPKSTREAM-AGGREGATE-KSTREAM-SOURCE-0000000001test
    Processor: KTABLE-TOSTREAM-0000000005 (stores: [])
      --> KSTREAM-SINK-0000000006
      <-- test
    Sink: KSTREAM-SINK-0000000006 (topic: output)
      <-- KTABLE-TOSTREAM-0000000005
```

Reply:
Alright, the I changed the tests so that they only use the correct number of builders. And I discovered that orElseGenerateWithPrefix exists. so that should fix these problems

Reply:
I still don't see that the test verifies the names? Again the question, why do we need to process data to verify if there right names are assigned?

Reply:
I added a test for this